### ISSUE TITLE
find and fix a bug in cascade

### ISSUE THREAD

**Issue by hanbingocean** [MIC-DKFZ/nnUNet#103]
Thank you for your excellent work and interest comments during code. ：）When use cascade net in kits, we found similar issue with https://github.com/MIC-DKFZ/nnUNet/issues/45,the performance of cascade net is significant lower (0.7)than low resolution net(0.86+). After some test, we find the reason is that the"segment from preview stage" is not do transpose by "transpose forward" before resize(I think probably resize is not use in fact) and concat with image. after modification, the code in predict_save_to_queue like:
# modify here
 seg_prev = seg_prev.transpose(transpose_forward)     
# finished
seg_reshaped = resize_segmentation(seg_prev, d.shape[1:], order=1, cval=0)                seg_reshaped = to_one_hot(seg_reshaped, classes)                d = np.vstack((d, seg_reshaped)).astype(np.float32)

After modification, the performance(0.88+) is better than low resolution net(0.86+)

**Comment by hanbingocean** [comment]
I forgot to say hello with you,Fabian.

**Comment by FabianIsensee** [comment]
Hi there :-)
Wow that is a hard one. I think KiTS is the only dataset I have that would produce this bug and since I did not use the cascade I did not notice.
Also back when I worked with KiTS nnU-Net did not have the transpose functionality so my cascade worked fine then.
I am currently working on a fix and will push it later today.
Best,
Fabian

**Comment by FabianIsensee** [comment]
Hi, I just pushed the changes. Can you verify if that works for you, too?
Best,
Fabian

**Comment by FabianIsensee** [comment]
(whoops I did not mean to close it yet)

**Comment by hanbingocean** [comment]
Hello Fabian,

I will test it tomorrow and tell you result after test.

**Comment by hanbingocean** [comment]
Hello Fabian,

I've tested it, it worked for me.
Thank you.
Please close it.

**Comment by yl255** [comment]
Hi, all
I updated the predict.py file.
But it did not work for me. :(

**Comment by FabianIsensee** [comment]
I have to thank you @hanbingocean! Without you that bug might have persisted for a while.

@yl255 can you please specify what exactly is not working? I recommend merging the current master into you code to make sure all changes are applied. Have you preprocessed the Kits data with the current version of the code? Maybe that was done with the old version that did not support transpose_forward?
Best,
Fabian

**Comment by yl255** [comment]
Hi,Fabian @FabianIsensee I will preprocess and train the data again. The data is not Kits.
My local code always keep the same as your's. 
And the 'transpose_forward'  is only added in the predict.py, so I just predicted the test data.
I do not know what is problem.
Thanks for your apply :)
Best,
Yaliang

**Comment by MengzhangLI** [comment]
Guten Morgen, Fabian:

Thanks for your excellent framework nnUNet! It really helps many researchers and people who are interested in MIA.

This bug was hard to troubleshoot and the problem is found and fixed just as @hanbingocean mentioned above. 
I've checked the new **predict.py** and it works out on KiTS Test predictions( https://kits19.grand-challenge.org/evaluation/results/ ). 

Results are shown below (mean Dice, Kidney Dice, Tumor Dice):
(1) cascade low + full (bugs **predict.py** as @hanbingocean mentioned above)
0.8202, 0.9364, 0.7041
(2) cascade low
0.8782, 0.9564, 0.7909
(3) cascade low + full( Our **predict.py** @hanbingocean fiexed)
0.8870, 0.9677, 0.8062
(4)cascade  low + full(Your new **predict.py** updated 6 days ago)
0.8870, 0.9677, 0.8062

Just as it shown above, (3) and (4) are literally identical so I think @yl255 problem is not your new codes' fault. We are also looking forward to following up @yl255 bug details and numerical results. 
Tips: A quick solution is **git clone** the lateset version and implementing your project again.

Plus: Cascade still have some potential bugs, for example, we cannot 
`python run/run_training.py 3d_cascade_fullres nnUNetTrainerCascadeFullRes TaskXX_MY_DATASET FOLD --ndet` 
on multiple GPUs (fold 0, 1, 2, 3, 4) simultaneously because each 3d_cascade procedure will do Write operation on output of last stage (**3d_lowres**), it may result in conflicts among each procedure. I understand it because I've read your annotations knowing you want to remind yourself handling history data ;).

We are looking forward to participating in and discussing this great framework. Hoping some new features or modules could be added in the future. 
Best,

**Comment by FabianIsensee** [comment]
Hi @MengzhangLI ,
thank you so much for the detailed analysis! Also for me the bug seems to be fixed. I actually used the cascade for a KiTS submission a few days ago - and the results are good.
How are you running the parallelization of the cascade on multiple GPUs? Most of the time it is as easy as adding an if statement:

```
if local_rank == 0:
   do_stuff()
```
The quickest way to run it in parallel would be to use pytorchs dataparallel implementation. But running nnU-Net this way actually changes a few things. For example dice loss computation over the entire batch will not work unless you hack it into the whole process. I have implementations for both DP and DDP un my personal repo and I will upload them soonish
Best,
Fabian

**Comment by MengzhangLI** [comment]
> Hi @MengzhangLI ,
> thank you so much for the detailed analysis! Also for me the bug seems to be fixed. I actually used the cascade for a KiTS submission a few days ago - and the results are good.
> How are you running the parallelization of the cascade on multiple GPUs? Most of the time it is as easy as adding an if statement:
> 
> ```
> if local_rank == 0:
>    do_stuff()
> ```
> 
> The quickest way to run it in parallel would be to use pytorchs dataparallel implementation. But running nnU-Net this way actually changes a few things. For example dice loss computation over the entire batch will not work unless you hack it into the whole process. I have implementations for both DP and DDP un my personal repo and I will upload them soonish
> Best,
> Fabian



> Hi @MengzhangLI ,
> thank you so much for the detailed analysis! Also for me the bug seems to be fixed. I actually used the cascade for a KiTS submission a few days ago - and the results are good.
> How are you running the parallelization of the cascade on multiple GPUs? Most of the time it is as easy as adding an if statement:
> 
> ```
> if local_rank == 0:
>    do_stuff()
> ```
> 
> The quickest way to run it in parallel would be to use pytorchs dataparallel implementation. But running nnU-Net this way actually changes a few things. For example dice loss computation over the entire batch will not work unless you hack it into the whole process. I have implementations for both DP and DDP un my personal repo and I will upload them soonish
> Best,
> Fabian

Guten Morgen, Fabian:
I just repeat commands in my server, like:

`CUDA_VISIBLE_DEVICES=X(0, 1, 2, etc.) python run/run_training.py 3d_cascade_fullres nnUNetTrainerCascadeFullRes TaskXX_MY_DATASET FOLD --ndet`

because at that time, I just wanted to reproduce cascade models numerical results ASAP. Each task is independent with each other thus I didn't use 
`if local_rank == 0:
   do_stuff()`

I never implemented multi-GPU experiments like you mentioned above because it will change some places in nnUNet (maybe some bugs hard to trouble shoot) and nobody is familiar with this framework like you ;) 

Thanks for your nice reply.

Best,

**Comment by FabianIsensee** [comment]
I guess I have misunderstood you. You are not running one model across several GPUs but instead running 4 models as the same time. Sorry for that :-D
Funny thing is I do that, too. But due to the configuration of our GPU cluster I have never encountered the problem you mentioned. That is because in our system each worker holds its own copy of the preprocessed data and thus the write conflict never happens. I will try to come up with a way to solve that problem.
I don't have an implementation for running the cascade in DP/DDP but the other trainers work well and as far as I can tell without bugs :-)
Best,
Fabian

**Comment by MengzhangLI** [comment]
B.T.W

Actually I've seen the newest result of KiTS. You've surpassed yourself successfully. ;)

Congratulations! 

Tumor segmentation is a hot and difficult topic because of highly imbalanced samples. I thought (1) proper data augmentation methods, (2) multi-stage(Low & High resolution task), (3) methods handling difficult samples like Gradient Harmonized or other novel loss functions.  

P.S. Would you like to disclose how you manage to do it? In MICCAI 2019 you said you would publish it in the future. I think many people are interested in it. 

Best,

**Comment by FabianIsensee** [comment]
Hi,
honestly, that was just nnU-Net V2, without manual intervention. The submission was the ensemble between 3d_fullres and 3d_cascade_fullres. On my CV, the single model was still worse than the residual UNet I submitted for the challenge, so no new innovation or anything. I think what gave the necessary boost was the ensembling which is something I did not do before.

I think what is needed most to tackle most segmentation problems is a proper datasets with high quality annotations. We spend so much time optimizing our models, wasting thousands of hours when all that is really needed (in most cases, not always!) is someone investing more time annotating and designing proper training sets. Especially nnU-net shows that already established concepts are really strong if implemented properly. Just have a look at all the crazy architectures people have been doing in the recent years and then nnUNet comes along with simple U-Nets and blows most of them away.
Given the current state of training sets, i am really sceptical about (3). 'difficult' samples migh just have been mislabeled and by emphasizing those you make your model overfit to something that is incorrect. If training sets were perfectly annotated this would be viable, but right now I am not sure if that doesn't actually harm your model.
(2) is only needed because current GPU memory limitations do not allow us to process patches, and thus receptive fields, large enough to capture sufficient contextual information
(1) what methods are you referring to exacltly? Yes there is a lot to be gained by data augmentation, and i would be interested in hearing what you have in mind

Best,
Fabian

**Comment by MengzhangLI** [comment]
Guten Morgen, Fabian:

Wow, I thought **ensemble** is really a pratical strategy for many competitions. 

nnUNet is robust and very successful, it would be a new benchmark in medical image segmentation field. Thanks for you and your group again! ;)

(1) I mentioned data augmentation because actually I implemented NoDA experiment and found dice of organ (0.97 -> 0.96) doesn't drop sharply like tumor (0.80 -> 0.70). I think operations in batchgenerators is enough in many cases. 

To be honest, I'm not familiar with data augmentation. Besides normal augmentation methods in batchgenerator, some people proposed deep learning frameworks to learn spatialand and apppearance transformation. There is a paper about medical image in CVPR 2019 **Data augmentation using learned transforms for one-shot medical image segmentation**  (https://arxiv.org/abs/1902.09383) the framework was based on 3D UNet, I hope it could be useful for you. Some other researchers use GAN to do it too.

(2) I really agree with you and actually I think 3D Unet to some degree is time-consuming and its acceleration is a promising field. 

(3) Annotation quality is absolutely important but medical image's annotation is difficult to achieve and generate even if for some professional doctors. I think many complicated deep learning models needs to prove themselves in many other datasets. 
However I think Gradient Harmonizing mechanism (https://arxiv.org/abs/1811.05181) might be useful in MIA because it forces models overlooked **outlier** samples which might be generated by wrong annotation. There is a naive follow-up paper in MICCAI 2019 (**Multi-Class Gradient Harmonized Dice Loss with Application to Knee MR Image Segmentation**), the result of numerical experiments in it is impressive.

Very happy to discuss and share my opinions with you, Fabian.

Best,

Mengzhang

**Comment by hanbingocean** [comment]
Now such issue has been fixed, so I close it.
